### ISSUE TITLE
Final is not used to narrow generic parameters

### ISSUE THREAD

**Issue by frogtd** [astral-sh/ty#3868]
### Summary

[Playground](https://play.ty.dev/88d22713-ce4c-40a6-8eae-a046025947c6)

```py
from dataclasses import dataclass
from typing import Generic, TypeVar, final

A = TypeVar("A", covariant=True)

class AlphaBound:
    pass

@final
class AlphaVariant(AlphaBound):
    pass

class BaseExpr(Generic[A]):
    pass

@dataclass
class Special(BaseExpr[AlphaVariant]):
    pass

def transform[B](expr: BaseExpr[B]) -> BaseExpr[B]:
    if isinstance(expr, Special):
        return Special()
    return expr
```

The above example should work, as in the `return Special()` branch, the check `isinstance(expr, Special)` had been done. 

Since `isinstance(expr, Special)` is true, this implies that `type(expr) <: Special` and then, since `Special <: BaseExpr[AlphaVariant]`, this implies `type(expr) <: BaseExpr[AlphaVariant]`. Since `type(expr) <: BaseExpr[B]` and `AlphaVariant` is final, this implies that `B = AlphaVariant`.

Then, as `Special <: BaseExpr[AlphaVariant] = BaseExpr[B]`, it can be returned.

Theoretically, this should also apply even without final, as a subclass of `Special` cannot re-subclass `BaseExpr` with a different generic parameter. However, if such a thing was implemented, it does mean that in the future if you could re-subclass a class with a different generic parameter, to e.g. narrow the type,  that would not be allowed to be supported by `ty`, since you would already implement analysis based on the lack of that feature.

**Comment by sharkdp** [comment]
> as in the `return Special()` branch, the check `isinstance(expr, Special)` had been done

I don't understand. The `isinstance` check does not affect the type of `Special()`? It only affects the type of `expr`.

> Since `isinstance(expr, Special)` is true, this implies that `type(expr) <: Special` and then, since `Special <: BaseExpr[AlphaVariant]`, this implies `type(expr) <: BaseExpr[AlphaVariant]`

Okay...

> Since `type(expr) <: BaseExpr[B]` and `AlphaVariant` is final

Here you jumped from `type(expr) <: BaseExpr[AlphaVariant]` to `type(expr) <: BaseExpr[B]`, but `B` is unconstrained.

> this implies that `B = AlphaVariant`

This is now how generics solvers work in Python type checkers. The body of a method does not contribute any constraints to how a type variable is solved. Type variables are solved for each individual *call* to a generic function. In the body of a function, the code needs to work for every possible solution.

> Then, as `Special <: BaseExpr[AlphaVariant] = BaseExpr[B]`, it can be returned.

No. Every type checker raises an error for your code.

**Comment by frogtd** [comment]
Just because every current typechecker returns an error does not mean that the code is wrong, and the point of the argument is showing that if `isinstance(expr, Special)` you can then return a different `Special`.

This does not constrain the type of `B` that could be passed into the function, but it shows if the check occurred, `B` must be a subtype of `AlphaVariant`.

**Comment by sharkdp** [comment]
Ok, I think I agree that this could be done *in principle*, but it is very unlikely that we will implement this. A simpler example for other readers would be something like:

```py
class IntList(list[int]): ...

def f[T](xs: list[T]) -> list[T]:
    if isinstance(xs, IntList):
        return IntList()
    return xs
```

where we would need to infer that inside `if isinstance(xs, IntList)`, `T` could really only be exactly `int`, and so it would be safe to return `IntList()`.

**Comment by sharkdp** [comment]
I'm tentatively closing this, but feel free to comment if anyone thinks it should be reopened.

**Comment by frogtd** [comment]
Note that that example you gave with `IntList` relies on the you only subclass once rule, as bool is a subclass of int.

**Comment by frogtd** [comment]
Ah, no it doesn't because list is invariant. That is a weaker analysis that seems more doable from an outsider's perspective.

**Comment by carljm** [comment]
#3988 independently requested this same feature. I agree it's quite low priority and I don't know if it would ever rise to the top of the implementation priority list, but since it seems clearly good in principle, I'm kind of inclined to reopen this with "wish" level priority.
